File: go/2022/day7.go
package _2022

import (
	"advent/utils"
	"fmt"
	"strings"
)

type Filesystem map[string]int

func cd(cwd, dirName string) string {
	switch dirName {
	case "..":
		parts := strings.Split(cwd, "/")
		cwd = strings.Join(parts[:len(parts)-1], "/")
	case "/":
		cwd = ""
	default:
		cwd = cwd + "/" + dirName
	}
	return cwd
}

func exploreFolders(input string) Filesystem {
	root := ""
	cwd := root
	contents := Filesystem{"/": 0}

	for _, line := range strings.Split(input, "\n") {
		if strings.HasPrefix(line, "$") {
			// parse commands
			cmd := strings.Replace(line, "$ ", "", -1)
			if strings.HasPrefix(cmd, "cd") {
				dirName := strings.Replace(cmd, "cd ", "", -1)
				cwd = cd(cwd, dirName)
			}
		} else {
			// parse ls contents
			if strings.HasPrefix(line, "dir") {
				dirName := strings.Replace(line, "dir ", "", -1)
				contents[cwd+"/"+dirName] = 0
			} else {
				split := strings.Fields(line)
				fileSize := split[0]
				fileName := split[1]
				fs := utils.ParseInt(fileSize)
				contents[cwd+"/"+fileName] = fs
			}
		}
	}
	return contents
}

func getFolderSize(contents Filesystem, folderPath string) int {
	total := 0
	for path, size := range contents {
		if strings.HasPrefix(path, folderPath) && !strings.HasSuffix(path, folderPath) {
			total += size
		}
	}
	return total
}

func Day7Part1(input string) string {
	contents := exploreFolders(input)
	total := 0
	for path, size := range contents {
		if size == 0 {
			folderSize := getFolderSize(contents, path)
			if folderSize < 100000 {
				total += folderSize
			}
		}
	}
	return fmt.Sprint(total)
}

func Day7Part2(input string) string {
	contents := exploreFolders(input)
	totalSpace := 70000000
	requiredSpace := 30000000
	occupiedSpace := getFolderSize(contents, "/")
	unusedSpace := totalSpace - occupiedSpace
	needToDelete := requiredSpace - unusedSpace

	allSizes := []int{}
	for path, size := range contents {
		if size == 0 {
			folderSize := getFolderSize(contents, path)
			allSizes = append(allSizes, folderSize)
		}
	}
	smallest := 9999999999
	for _, folderSize := range allSizes {
		if folderSize >= needToDelete && folderSize < smallest {
			smallest = folderSize
		}
	}
	return fmt.Sprint(smallest)
}

File: python/puzzles/y2022/day7.py
import utils

example = """$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k"""


def explore_folders(raw: str) -> dict[str:int]:
    root = cwd = ""
    contents = {"/": 0}
    for line in raw.split("\n"):
        if line.startswith("$"):
            _, cmd = line.split(" ", maxsplit=1)
            match cmd.split():
                case ["cd", dir_name]:
                    match dir_name:
                        case "..":
                            cwd, _ = cwd.rsplit("/", maxsplit=1)
                        case "/":
                            cwd = root
                        case _:
                            cwd = "/".join([cwd, dir_name])

        else:  # must be lines following "ls"
            match line.split():
                case ["dir", dir_name]:
                    contents["/".join([cwd, dir_name])] = 0
                case [file_size, file_name]:
                    contents["/".join([cwd, file_name])] = int(file_size)
    return contents


def get_folder_size(contents: dict[str:int], folder_path: str) -> int:
    return sum(
        size
        for path, size in contents.items()
        if path.startswith(folder_path) and not path.endswith(folder_path)
    )


@utils.profile
def part1(raw: str) -> int:
    contents = explore_folders(raw)
    folders = {k for k, v in contents.items() if v == 0}
    total = 0
    for folder in folders:
        if (size := get_folder_size(contents, folder)) < 100000:
            total += size
    return total


@utils.profile
def part2(raw: str) -> int:
    contents = explore_folders(raw)
    TOTAL_SPACE = 70_000_000
    REQUIRED_SPACE = 30_000_000
    occupied_space = get_folder_size(contents, folder_path="/")
    unused_space = TOTAL_SPACE - occupied_space
    need_to_delete = REQUIRED_SPACE - unused_space

    folders = {k for k, v in contents.items() if v == 0}
    all_sizes = [get_folder_size(contents, f) for f in folders]
    return min(s for s in all_sizes if s > need_to_delete)


if __name__ == "__main__":
    raw = utils.load_puzzle_input("2022/day7")
    assert part1(raw) == 1792222
    assert part2(raw) == 1112963
